packages/@aws-cdk/aws-appmesh/lib/virtual-node-listener.ts

```diff
@@ -0,0 +1,348 @@
+          protocol: this.protocol,
+        },
+        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
+        timeout: this.timeout ? this.renderTimeout(this.timeout) : undefined,
+      },
+    };
+  }
+}
+
+/**
+* Represents the properties needed to define an HTTP2 Listener for a VirtualGateway
+*/
+class Http2NodeListener extends HttpNodeListener {
+  constructor(props: HttpNodeListenerProps = {}) {
+    super(props);
+    this.protocol = Protocol.HTTP2;
+  }
+}
+
+/**
+ * Represents the properties required to define a GRPC Listener for a VirtualNode.
+ */
+class GrpcNodeListener extends VirtualNodeListener {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Listener timeout property for GRPC protocol
+   *
+   * @default - none
+   */
+  readonly timeout?: GrpcTimeout;
+
+  /**
+   * Listener timeout property for GRPC protocol
+   *
+   * @default - none
+   */
+  protected protocol: Protocol = Protocol.GRPC;
+
+  constructor(props: GrpcNodeListenerProps = {}) {
+    super();
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+    this.timeout = props.timeout;
+  }
+
+  /**
+   * Return Listener for GRPC protocol when Listener is added to Virtual Node.
+   */
+  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
+    return {
+      listener: {
+        portMapping: {
+          port: this.port,
+          protocol: Protocol.GRPC,
+        },
+        healthCheck: this.healthCheck ? this.renderHealthCheck(this.healthCheck) : undefined,
+        timeout: this.timeout ? this.renderTimeout(this.timeout) : undefined,
+      },
+    };
+  }
+}
+
+/**
+ * Represents the properties required to define a TCP Listener for a VirtualNode.
+ */
+class TcpNodeListener extends VirtualNodeListener {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Listener timeout property for TCP protocol
+   *
+   * @default - none
+   */
+  readonly timeout?: TcpTimeout;
+
+  /**
+   * Listener timeout property for HTTP protocol
+   *
+   * @default - none
+   */
+  protected protocol: Protocol = Protocol.TCP;
+
+  constructor(props: TcpNodeListenerProps = {}) {
+    super();
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+    this.timeout = props.timeout;
+  }
+
+  /**
+   * Return Listener for TCP protocol when Listener is added to Virtual Node.
+  */
+  public bind(_scope: cdk.Construct): VirtualNodeListenerConfig {
```

Comment:
All of these `bind`s are exactly the same... can you move this implementation up to `VirtualNodeListener`, and use `abstract` properties there like `protocol` to get rid of this duplication?

Reply:
The timeout property is not generic for a VirtualNodeListener. i.e. the timeout property for GRPC protocol would be of the type `GrpcTimeout` whereas for TCP protocol it is `TcpTimeout`. The `TcpTimeout` and `GrpcTimeout` are different as they have different properties.
```
export interface GrpcTimeout {

  readonly idle?: cdk.Duration;

  readonly perRequest?: cdk.Duration;
}

export interface TcpTimeout {

  readonly idle?: cdk.Duration;
}
```
Even though the bind are exactly the same, the timeout property would make them different. Therefore `bind` methods was implemented for each `*NodeListener` class.

Reply:
Then do this:

```ts
// module-private
interface Timeout {
  readonly idle?: cdk.Duration;

  readonly perRequest?: cdk.Duration;
}
```

And use `Timeout` as the type in `VirtualNodeListener`.

Because of TypeScript's structural typing, you can pass instances of `GrpcTimeout` and `TcpTimeout` where `Timeout` is expected.